core/src/main/scala/kafka/controller/ReplicaStateMachineV2.scala

```diff
@@ -0,0 +1,334 @@
+    configResponses.foreach { configResponse =>
+      val partition = configResponse.ctx.asInstanceOf[TopicAndPartition]
+      if (Code.get(configResponse.rc) == Code.OK) {
+        val overrides = ConfigEntityZNode.decode(configResponse.data)
+        val logConfig = LogConfig.fromProps(config, overrides.getOrElse(new Properties))
+        val newLeader = candidateLeaderAndIsrs(partition).leader
+        val adjustedIsr = if (!logConfig.uncleanLeaderElectionEnable) currentLeaderAndIsrs(partition).isr else candidateLeaderAndIsrs(partition).isr
+        val adjustedLeaderAndIsr = currentLeaderAndIsrs(partition).newLeaderAndIsr(newLeader, adjustedIsr)
+        adjustedLeaderAndIsrs.put(partition, adjustedLeaderAndIsr)
+      } else if (Code.get(configResponse.rc) == Code.NONODE) {
+        val logConfig = LogConfig.fromProps(config, new Properties)
+        val newLeader = candidateLeaderAndIsrs(partition).leader
+        val adjustedIsr = if (!logConfig.uncleanLeaderElectionEnable) currentLeaderAndIsrs(partition).isr else candidateLeaderAndIsrs(partition).isr
+        val adjustedLeaderAndIsr = currentLeaderAndIsrs(partition).newLeaderAndIsr(newLeader, adjustedIsr)
+        adjustedLeaderAndIsrs.put(partition, adjustedLeaderAndIsr)
+      } else {
+        failed.put(partition, KeeperException.create(Code.get(configResponse.rc)))
+      }
+    }
+    (adjustedLeaderAndIsrs.toMap, failed.toMap)
+  }
+
+  private def updateLeaderAndIsrToZk(leaderAndIsrs: Map[TopicAndPartition, LeaderAndIsr]):
+  (Map[TopicAndPartition, LeaderAndIsr],
+    Seq[TopicAndPartition]) = {
+    val successfulUpdates = mutable.Map.empty[TopicAndPartition, LeaderAndIsr]
+    val badVersions = mutable.Buffer.empty[TopicAndPartition]
+    val updatesToRetry = mutable.Buffer.empty[TopicAndPartition]
+    val leaderIsrAndControllerEpochAndContexts = leaderAndIsrs.map { case (partition, leaderAndIsr) => partition -> (LeaderIsrAndControllerEpoch(leaderAndIsr, controller.epoch), partition) }
+    val setDataResponses = try {
+      KafkaControllerZkUtils.setTopicPartitionStates(zookeeperClient, leaderIsrAndControllerEpochAndContexts)
+    } catch {
+      case _: Exception =>
+        leaderAndIsrs.keys.foreach(partition => updatesToRetry += partition)
+        return (successfulUpdates.toMap, updatesToRetry)
+    }
+    setDataResponses.foreach { setDataResponse =>
+      val partition = setDataResponse.ctx.asInstanceOf[TopicAndPartition]
+      if (Code.get(setDataResponse.rc) == Code.OK) {
+        val updatedLeaderAndIsr = leaderAndIsrs(partition).withZkVersion(setDataResponse.stat.getVersion)
+        successfulUpdates.put(partition, updatedLeaderAndIsr)
+      } else if (Code.get(setDataResponse.rc) == Code.BADVERSION) {
+        badVersions += partition
+      } else {
+        updatesToRetry += partition
+      }
+    }
+    val getDataResponses = try {
+      KafkaControllerZkUtils.getTopicPartitionStates(zookeeperClient, badVersions.zip(badVersions).toMap)
```

Comment:
It attempts to mimic the existing behavior in ReplicationUtils.checkLeaderAndIsrZkData (which is called from ReplicationUtils.updateLeaderAndIsr).

Strictly speaking I don't think this logic is required. I think it just attempts to make progress on a version conflict (for instance if the partition leader concurrently updated isr) instead of starting over.

Reply:
Ok, that logic is just to handle the possibility that a previous write has succeeded on a ConnectionLossException. It would be useful to document that.